core/services/keeper/registry_synchronizer_process_logs.go

```diff
@@ -0,0 +1,195 @@
+package keeper
+
+import (
+	"sync"
+
+	"github.com/smartcontractkit/chainlink/core/internal/gethwrappers/generated/keeper_registry_wrapper"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/log"
+	"github.com/smartcontractkit/chainlink/core/services/postgres"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+)
+
+func (rs *RegistrySynchronizer) processLogs(head models.Head) {
+	wg := sync.WaitGroup{}
+	wg.Add(4)
+	go rs.handleSyncRegistryLogs(head, wg.Done)
+	go rs.handleUpkeepCanceledLogs(head, wg.Done)
+	go rs.handleUpkeepRegisteredLogs(head, wg.Done)
+	go rs.handleUpkeepPerformedLogs(head, wg.Done)
+	wg.Wait()
+}
+
+func (rs *RegistrySynchronizer) handleSyncRegistryLogs(head models.Head, done func()) {
+	defer done()
+	oldEnough := isOldEnoughConstructor(head, rs.minConfirmations)
+	for {
+		i := rs.mailRoom.mbSyncRegistry.RetrieveIf(oldEnough)
+		if i == nil {
+			return
+		}
+		broadcast, ok := i.(log.Broadcast)
+		if !ok {
+			logger.Errorf("RegistrySynchronizer: invariant violation, expected log.Broadcast but got %T", broadcast)
+			continue
+		}
+		was, err := broadcast.WasAlreadyConsumed()
+		if err != nil {
+			logger.Warn(err)
+			continue
+		}
+		if was {
+			continue
+		}
+		if !head.IsInChain(broadcast.RawLog().BlockHash) {
+			continue
+		}
+		_, err = rs.syncRegistry()
+		if err != nil {
+			logger.Error(err)
+			continue
+		}
+		err = broadcast.MarkConsumed()
+		logger.ErrorIf(err)
+	}
+}
+
+func (rs *RegistrySynchronizer) handleUpkeepCanceledLogs(head models.Head, done func()) {
+	defer done()
+	oldEnough := isOldEnoughConstructor(head, rs.minConfirmations)
+	for {
+		i := rs.mailRoom.mbUpkeepCanceled.RetrieveIf(oldEnough)
+		if i == nil {
+			return
+		}
+		broadcast, ok := i.(log.Broadcast)
+		if !ok {
+			logger.Errorf("RegistrySynchronizer: invariant violation, expected log.Broadcast but got %T", broadcast)
+			continue
+		}
+		was, err := broadcast.WasAlreadyConsumed()
```

Comment:
I've seen this sort of logic repeated a lot. I wonder if we want to pull some of it upwards, or perhaps if the interface can be simplified?

Reply:
I agree this isn't dry at all. I think there's a bit of a dry / readable tradeoff here, I'm worried if I try to hard to reuse code between these methods it will look more confusing than it already is... plus this should be drastically simpler once the logBroadcaster's confirmations feature is finished. So I'm inclined to leave this less DRY, and clean it up once we have the LB ready to roll.

Reply:
yeah i'll see about creating some mailbox helpers on my branch